drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1515 @@
+                                                &AcceptSocketDispatch);
+
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ListenDispatch->WskAcceptEvent returned non-successful status 0x%08x\n", Status);
+        }
+        else
+        {
+            memcpy(&AcceptSocket->RemoteAddress, RemoteAddress, sizeof(AcceptSocket->RemoteAddress));
+        }
+    }
+    else
+    {
+        SocketPut(AcceptSocket);	/* close the AcceptSocket */
+    }
+    ListenSocket->ListenIrp = NULL;
+
+    SocketPut(AcceptSocket);
+    SocketPut(ListenSocket);
+
+    ExFreePoolWithTag(ListenContext->ReturnConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    ExFreePoolWithTag(ListenContext->RequestConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+    ExFreePoolWithTag(ListenContext, TAG_NETIO);
+
+    /* And wait for the next incoming connection. */
+    /* This is done in a separate thread at IRQL = 0 */
+
+    QueueListening(ListenSocket);
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS
+StartListening(_In_ PWSK_SOCKET_INTERNAL ListenSocket)
+{
+    PIRP tdiIrp;
+    NTSTATUS status;
+    PLISTEN_CONTEXT lc;
+    PWSK_SOCKET_INTERNAL AcceptSocket;
+
+    FUNCTION_TRACE;
+
+    if (ListenSocket->LocalAddressHandle == NULL)
+    {
+        DPRINT1("LocalAddressHandle is not set, need to bind your socket before listening\n");
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    status = CreateSocket(ListenSocket->family, ListenSocket->type, ListenSocket->proto, WSK_FLAG_CONNECTION_SOCKET, &AcceptSocket);
+    if (!NT_SUCCESS(status))
+    {
+        DPRINT1("Could not create AcceptSocket, status is 0x%08x\n", status);
+        return status;
+    }
+    AcceptSocket->ListenSocket = ListenSocket;
+    /* Put when the AcceptSocket is closed */
+    SocketGet(AcceptSocket->ListenSocket);
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    lc = ExAllocatePoolUninitialized(NonPagedPool, sizeof(*lc), TAG_NETIO);
+    if (lc == NULL)
+    {
+        goto err_out_free_accept_socket;
+    }
+    lc->ListenSocket = ListenSocket;
+    lc->AcceptSocket = AcceptSocket;
+
+    tdiIrp = NULL;
+
+    status = TdiAssociateAddressFile(ListenSocket->LocalAddressHandle, AcceptSocket->ConnectionFile);
+    if (!NT_SUCCESS(status))
+    {
+        goto err_out_free_lc;
+    }
+    AcceptSocket->ConnectionFileAssociated = TRUE;
+
+    lc->RequestConnectionInfo = NULL;
+    TdiBuildNullConnectionInfo(&lc->RequestConnectionInfo, TDI_ADDRESS_TYPE_IP);
+    if (lc->RequestConnectionInfo == NULL)
+    {
+        goto err_out_free_lc_and_disassociate;
+    }
+    lc->RequestConnectionInfo->OptionsLength = 0;
+
+    lc->ReturnConnectionInfo = NULL;
+    TdiBuildNullConnectionInfo(&lc->ReturnConnectionInfo, TDI_ADDRESS_TYPE_IP);
+    if (lc->ReturnConnectionInfo == NULL)
+    {
+        goto err_out_free_lc_and_req_conn_info;
+    }
+    lc->ReturnConnectionInfo->OptionsLength = 0;
+
+    SocketGet(ListenSocket);
```

Comment:
Why is it necessary to "get" again the ListenSocket? it appears it's already done at line 567 when you do:
```c
    AcceptSocket->ListenSocket = ListenSocket;
    /* Put when the AcceptSocket is closed */
    SocketGet(AcceptSocket->ListenSocket);
```